bin/tx-decoder/src/decoder.rs

```diff
@@ -0,0 +1,66 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of substrate-desub.
+//
+// substrate-desub is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// substrate-desub is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with substrate-desub.  If not, see <http://www.gnu.org/licenses/>.
+
+
+use desub::{decoder::Decoder as DecoderOld, TypeDetective};
+use desub_v14::{Decoder as DecoderNew, Metadata as MetadataNew};
+use desub::decoder::Chain;
+use std::{collections::HashMap, convert::TryInto};
+use anyhow::{Error, anyhow};
+
+use crate::app::SpecVersion;
+
+pub struct Decoder {
+	old: DecoderOld,
+	new: HashMap<u32, DecoderNew>,
+}
+
+impl Decoder {
+	pub fn new(types: impl TypeDetective + 'static, chain: Chain) -> Self {
+		Self { old: DecoderOld::new(types, chain), new: HashMap::new() }
+	}
+
+	pub fn register_version(&mut self, version: SpecVersion, meta: &[u8]) -> Result<(), Error> {
+		let new = MetadataNew::from_bytes(meta);
+		if new.is_err() {
+			self.old.register_version(version.try_into()?, meta);
+		} else {
+			self.new.insert(version.try_into()?, DecoderNew::with_metadata(new?));
+		};
+		Ok(())
+	}
+
+	// Decodes extrinsics and serializes to String
+	pub fn decode_extrinsics(&self, version: SpecVersion, data: &[u8]) -> Result<String, Error> {
+		if self.is_version_new(version) {
+			let decoder = self.new.get(&version.try_into()?).ok_or_else(|| anyhow!("version {} not found for new decoder", version))?;
+			let ext = decoder.decode_extrinsics(data)?;
+			// Ok(serde_json::to_string_pretty(&ext)?)
```

Comment:
Make it a `trace` log?

Reply:
Ah yeah, it is. I actually just left it here to remind me to uncomment it at some point; the output of `decode_extrinsics` in v14 doesn't impl `Serialize/Deserialize` yet